File: web/lib/build_url.dart
library build_url;
import 'package:polymer/polymer.dart';
import 'models.dart';

@CustomTag('build-url')
class BuildURL extends PolymerElement with ObservableMixin {
  static const List<int> _asciipoints = const [
    48,49,50,51,52,53,54,55,56,57, // 0-9
    65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90, // A-Z
    97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122, // a-z
    45,95 // -_
    ];
  static final Map<int, String> _ALPHABET = new Map.fromIterable(
      new Iterable.generate(64, (i) => i),
      key: (int n) => n,
      value: (int n) => new String.fromCharCode(_asciipoints[n])
      );
  static final Map<String, int> _ALPHABET_REVERSE = new Map.fromIterable(
      new Iterable.generate(64, (i) => i),
      key: (int n) => new String.fromCharCode(_asciipoints[n]),
      value: (int n) => n
      );
  @observable String value = '';
  bool get applyAuthorStyles => true;

  String _hash(int num) {
    StringBuffer result = new StringBuffer();
    while (num != 0) {
      result.write(_ALPHABET[num&63]);
      num = num>>6;
    }
    return result.toString();
  }

  int _unhash(String msg) {
    int result = 0;
    for (String c in msg.split('')) {
      result = (result<<6) | _ALPHABET_REVERSE[c];
    }
    return result;
  }

  String hash_build(Job job, int level, List<SkillLevel> slevels) {
    StringBuffer result = new StringBuffer();
    for (SkillLevel slevel in slevels) {
      String tmp = _hash( (slevel.skill.job.index<<10)|(slevel.level<<5)|slevel.skill.tree_index );
      result.write( tmp.length==2 ? tmp : '0${tmp}');
    }
    result.write('.${_hash((job.id<<7)|level)}');
    return result.toString();
  }

  List<SkillLevel> unhash_slevels(List<Job> job_byindx, String msg){
    List<SkillLevel> result = new List<SkillLevel>();
    String skillmsg = msg.split('.')[0];
    for (int n=0; n < skillmsg.length; n+=2) {
      int skillinfo = _unhash(skillmsg.substring(n, n+2));
      int job_indx = skillinfo>>10;
      int level = (skillinfo>>5)&31;
      int tree_index = skillinfo&31;
      try {
        SkillLevel slevel = job_byindx[job_indx].skilltree[tree_index].level[level];
        result.add(slevel);
      } catch (e) {
        continue;
      }
    }
  }
}

File: polymer-elements/build_url.js
Polymer('build-url', {
    value: '',

    created: function() {//{{{
        console.debug("build-url.created - enter");
        this.alphabet = {};
        this.alphabet_reverse = {};

        var n = 0;
        for (var i=45; i<123; i++) {
            if (i == 46){ i = 48; }
            if (i == 58){ i = 65; }
            if (i == 91){ i = 95; }
            if (i == 96){ i = 97; }
            this.alphabet[n] = String.fromCharCode(i);
            this.alphabet_reverse[String.fromCharCode(i)] = n;
            n++;
        }
        console.debug("build-url.created - exit");
    },//}}}

    hash: function(num, len) {//{{{
        console.debug("build-url.hash - enter");
        var result = '';
        while (num !== 0) {
            result += this.alphabet[num&63];
            num = num>>6;
        }
        if (result.length < len) { 
            result = result + new Array(1+len-result.length).join(this.alphabet[0]);
        }
        console.debug("build-url.hash - exit");
        return result;
    },//}}}

    unhash: function(msg) {//{{{
        console.debug("build-url.unhash - enter");
        var result = 0;
        for (var n=msg.length-1; n>=0; n--) {
            result = (result<<6) | this.alphabet_reverse[msg[n]];
        }
        console.debug("build-url.unhash - exit");
        return result;
    },//}}}

    hash_build: function() {//{{{
        console.debug("build-url.hash_build - enter");
        var n, i, slevel, skill;
        var result = '';
        var tmp = 0;
        var tmp_count = 0;
        for (n in model.job_byindx){
            for (i=0; i<24; i++){
                try {
                    skill = model.job_byindx[n].skill[i];
                    slevel = main.skill_grid[n].icon[skill.id].level;
                } catch (e) {
                    slevel = 0;
                }
                tmp = (tmp<<5)|slevel;
                tmp_count++;
                if (tmp_count == 6){
                    result += this.hash(tmp, 5);
                    tmp = 0;
                    tmp_count = 0;
                }
            }
        }
        console.debug("build-url.hash_build - exit");
        return result;
    },//}}}

    hash_job: function(job, level) {//{{{
        console.debug("build-url.hash_job - enter");
        var hash = this.hash(job.id<<7|level);
        console.debug("build-url.hash_job - exit");
        return hash;
    },//}}}

    unhash_build: function(message) {//{{{
        console.debug("build-url.unhash_build - enter");
        var n, i, level, skill, slevel;
        var msg = message.split('.')[0];
        var result = {};
        var num = 0;
        var count = 0;

        for (n in model.job_byindx[0].skill) {
            skill = model.job_byindx[0].skill[n];
            slevel = skill.level[1];
            if (slevel.req_level == 1 && slevel.sp_cost === 0) {
                result[skill.id] = 1;
            }
        }
        for (n in model.job_byindx){
            for (i=0; i<24; i++) {
                if (count%6 === 0) {
                    num = this.unhash(msg.slice(0,5));
                    msg = msg.slice(5);
                }
                level = num>>(25-(count%6)*5)&31;
                if (level>0) { result[model.job_byindx[n].skill[i].id] = level;}
                count++;
            }
        }
        console.debug("build-url.unhash_build - exit");
        return result;
    },//}}}

    unhash_job: function(message) {//{{{
        console.debug("build-url.unhash_job - enter");
        var msg = this.unhash(message.split('.')[1]);
        var result = { 'job':msg>>7 , 'level':msg&127 };
        console.debug("build-url.unhash_job - exit");
        return result;
    }//}}}

});
